lib/puppet/application/lookup.rb

```diff
@@ -331,6 +331,25 @@ def generate_scope
     end
 
     node = Puppet::Node.indirection.find(node) unless node.is_a?(Puppet::Node) # to allow unit tests to pass a node instance
+
+    if options[:fact_file]
+      original_facts = node.facts.values
+      if options[:fact_file].end_with?("json")
+        given_facts = JSON.parse(File.read(options[:fact_file]))
+      else
+        given_facts = YAML.load(File.read(options[:fact_file]))
+
+        # If the data is not correctly formatted this will be a string so we should warn the user
+        # Otherwise both JSON and YAML will provide sane parse errors for the user if the data is
+        # incorrectly formatted
+        if !given_facts.instance_of?(Hash)
```

Comment:
This check is applied when using YAML only. This check is needed for JSON as well (the parsed result cannot be a String when parsing JSON but it might be an Array). So move the check out of this `if`